runtime/lua/vim/semver.lua

```diff
@@ -0,0 +1,263 @@
+local function extract_components_strict(version)
+  local major, minor, patch, prerelease_and_build = version:match('^v?(%d+)%.(%d+)%.(%d+)(.*)$')
+  return tonumber(major), tonumber(minor), tonumber(patch), prerelease_and_build
+end
+
+---@private
+--- Extracts the major, minor, patch and preprelease and build components from
+--- `version`. When `minor` and `patch` components are not found (nil), coerce
+--- them to 0.
+---@param version string Version string
+local function extract_components_loose(version)
+  local major, minor, patch, prerelease_and_build = version:match('^v?(%d+)%.?(%d*)%.?(%d*)(.*)$')
+  major = tonumber(major)
+  minor = tonumber(minor) or 0
+  patch = tonumber(patch) or 0
+  return major, minor, patch, prerelease_and_build
+end
+
+---@private
+--- Validates the prerelease and build string e.g. "-rc1+build.0". If the
+--- prerelease, build or both are valid forms then it will return true, if it
+--- is not of any valid form, it will return false.
+---@param prerelease_and_build string
+---@return boolean
+local function is_prerelease_and_build_valid(prerelease_and_build)
+  local has_build = prerelease_and_build:match('^%+[%w%.]+$')
+  local has_prerelease = prerelease_and_build:match('^%-[%w%.]+$')
+  local has_prerelease_and_build = prerelease_and_build:match('^%-[%w%.]+%+[%w%.]+')
+  return has_build or has_prerelease or has_prerelease_and_build
+end
+
+---@private
+---@param prerelease_and_build string
+---@return boolean
+local function has_prerelease_and_build(prerelease_and_build)
+  return prerelease_and_build ~= nil and prerelease_and_build ~= ''
+end
+
+---@private
+---@param version string
+---@return string
+local function create_err_msg(version)
+  return string.format('%s is not a valid version', version)
+end
+
+--- Parses a semantically formatted version string into a table.
+---
+--- Supports leading "v" and leading and trailing whitespace in the version
+--- string. e.g. `" v1.0.1-rc1+build.2"` , `"1.0.1-rc1+build.2"`, `"v1.0.1-rc1+build.2"`
+--- and `"v1.0.1-rc1+build.2 "` will be parsed as:
+---
+--- `{ major = 1, minor = 0, patch = 1, prerelease = 'rc1', build = 'build.2' }`
+---
+---@param version string Version string to be parsed.
+---@param opts table|nil Optional keyword arguments:
+---                      - strict (boolean):  when set to `true` an error will be thrown for version
+---                      strings which are not conforming to semver sepecifications (v2.0.0), (see
+---                      semver.org/spec/v2.0.0.html for details), this means that
+---                      `semver.parse('v1.2)` will throw an error. When set to `false`,
+---                      `semver.parse('v1.2)` will coerce 'v1.2' to 'v1.2.0' and return the table:
+---                      `{ major = 1, minor = 2, patch = 0 }`. Defaults to false.
+---@return table parsed_version Parsed version table
+function M.parse(version, opts)
+  if type(version) ~= 'string' then
+    error(create_err_msg(version))
```

Comment:
error makes sense here, because this is like a "syntax" error (validation).

For other cases below, where the version string wasn't parseable for whatever reason, it seems useful to return `nil` instead of raising an error. Then the caller can use parse() as a "validate" function. That addresses this request:

> would it be possible to also include a vim.semver.validate function that just returns true or false if the string is a valid semver?